### DISCUSSION TITLE
How to get the border router ip address from an end device

### DISCUSSION THREAD

**Discussion by 10andy0** [openthread/openthread#9811] [Q&A] [answered]
Hi there,

   Is there an API (CLI) to get list of border routers in the network form an end device?
I saw some articles where DNS-SD and SRP service discovery protocols are suggested.

With DNS-SD could not find any examples on how to query for _meshcop._udp service from an end device (FTD or MTD),
only examples on how to do it from Linux side.

With SRP I had to know an border router IP address in order to get it working (a chicken and egg situation)

On the border router I added coap-srv service, as follows:
```
> srp client host name my-br
Done
> srp client host address auto
Done
> srp client service add coap-srv _ipps._udp 12345
Done
> srp client autostart enable
Done
> srp server host
my-br.default.service.arpa.
    deleted: false
    addresses: [xxxxx]
```
Then I had to run 
> ipaddr mleid
yyyyy

to get the IP address to use it on an MTD node, as follows:
```
> dns config yyyyy
Done
> dns service coap-srv _ipps._udp.default.service.arpa.
DNS service resolution response for coap-srv for service _ipps._udp.default.service.arpa.
Port:12345, Priority:0, Weight:0, TTL:6494
Host:my-br.default.service.arpa.
HostAddress:xxxxx TTL:6494
TXT:[] TTL:6494
```

It seems to me that I am doing something wrong here. Any suggestions are welcome. 

Cheers,
Andy

**Comment by wgtdkp** [comment]
I think you can use `> dns service _meshcop._udp.default.service.arpa.`. The host address should be the address of the infrastructure interface on the Border Router, but can still reach to that address from a Thread end device

**Reply by 10andy0** [reply]
Thank you for your reply, but this doesn't work:
```
> dns service _meshcop._udp.default.service.arpa.
Error 7: InvalidArgs
```

**Reply by 10andy0** [reply]
Let me add some practical example here. Let's say there is some service running on the border router, say CoAP server and I would like to connect to the server from an end device in the Thread network. How can I connect from the device to the server without IP address hardcoding?  Reference to any examples, articles, documentation, etc. is appreciated.

**Reply by wgtdkp** [reply]
Sorry, you should use
```
dns browse _meshcop._udp.default.service.arpa.
```
to discover available BRs

then
```
dns service <service-instance-name> _meshcop._udp.default.service.arpa.
```

"service-instance-name" should be the name returned in the "dns browse" command

> Let me add some practical example here. Let's say there is some service running on the border router, say CoAP server and I would like to connect to the server from an end device in the Thread network. How can I connect from the device to the server without IP address hardcoding? Reference to any examples, articles, documentation, etc. is appreciated.


Even with the above "_mshcop._udp" works, I think the idea of registering a specific "coap-srv" of your own type is the best approach - so that you can add your own service configuration info in the TXT data

**Reply by 10andy0** [reply]
That's ok.

```
> dns browse _meshcop._udp.default.service.arpa.
DNS browse response for _meshcop._udp.default.service.arpa.
Error 28: ResponseTimeout

```
it works only when prior to that `> dns config yyy` was run. Where yyy is a boarder router ip address.

**Comment by abtink** [comment, marked as answer]
> With SRP I had to know an border router IP address in order to get it working (a chicken and egg situation)

The `AutoEnable` APIs in SRP client will handle this:

https://github.com/openthread/openthread/blob/c73829fdbf5b0019dc611e913674021492c3fa9e/include/openthread/srp_client.h#L271-L310

You can also get the address of the currently selectd server from client:
https://github.com/openthread/openthread/blob/c73829fdbf5b0019dc611e913674021492c3fa9e/include/openthread/srp_client.h#L245-L256

BRs that provide SRP server (or DNS resolver/Adv-proxy) function will publish specific "SRP/DNS service entries" in Network Data. This is parsed and monirtoed by SRP client and/or DNS client.

---

A related config for how DNS client will pick the resolver/server address:
https://github.com/openthread/openthread/blob/c73829fdbf5b0019dc611e913674021492c3fa9e/src/core/config/dns_client.h#L93-L107

**Reply by 10andy0** [reply]
Thank you Abtin! This is exactly what I was looking for!

**Comment by 10andy0** [comment]
Resolved!
